src/w3d/renderer/linegroup.cpp

```diff
@@ -0,0 +1,358 @@
+        } else {
+            m_shader.Set_Texturing(ShaderClass::TEXTURING_DISABLE);
+        }
+
+        VertexMaterialClass *material = VertexMaterialClass::Get_Preset(VertexMaterialClass::PRELIT_DIFFUSE);
+
+        DX8Wrapper::Set_Material(material);
+        DX8Wrapper::Set_Shader(m_shader);
+        DX8Wrapper::Set_Texture(0, m_texture);
+
+        if (material != nullptr) {
+            material->Release_Ref();
+        }
+
+        captainslog_assert(m_startLoc != nullptr && m_startLoc->Get_Array());
+        captainslog_assert(m_endLoc != nullptr && m_endLoc->Get_Array());
+
+        const bool sort = m_shader.Get_Dst_Blend_Func() != ShaderClass::DSTBLEND_ZERO
+            && m_shader.Get_Alpha_Test() == ShaderClass::ALPHATEST_DISABLE && W3D::Is_Sorting_Enabled();
+
+        static Vector3 _offset_a(GameMath::Cos(DEG_TO_RAD(90)), GameMath::Sin(DEG_TO_RAD(90)), 0);
+        static Vector3 _offset_b(GameMath::Cos(DEG_TO_RAD(210)), GameMath::Sin(DEG_TO_RAD(210)), 0);
+        static Vector3 _offset_c(GameMath::Cos(DEG_TO_RAD(330)), GameMath::Sin(DEG_TO_RAD(330)), 0);
+        static Vector3 _offset[3];
+
+        _offset[0].Set(_offset_a);
+        _offset[1].Set(_offset_b);
+        _offset[2].Set(_offset_c);
+
+        Matrix4 old_view;
+        DX8Wrapper::Get_Transform(D3DTS_VIEW, old_view);
+
+        Matrix4 identity(true);
+        DX8Wrapper::Set_Transform(D3DTS_WORLD, identity);
+
+        if (Get_Flag(TRANSFORM)) {
+            Matrix3D model_view;
+
+            model_view = rinfo.m_camera.Get_Transform();
+            model_view.Set_Translation(Vector3(0, 0, 0));
+            model_view.Get_Orthogonal_Inverse(model_view);
+
+            for (int i = 0; i < 3; ++i) {
+                Matrix3D::Transform_Vector(model_view, _offset[i], &_offset[i]);
+            }
+        } else {
+            DX8Wrapper::Set_Transform(D3DTS_VIEW, identity);
+        }
+
+        int polygon_count = 0;
+        int index_count = 0;
+        int vertex_count = 0;
+
+        if (m_mode == TETRAHEDRON) {
+            polygon_count = 4 * m_lineCount;
+            index_count = 3 * polygon_count;
+            vertex_count = 4 * m_lineCount;
+        } else if (m_mode == PRISM) {
+            polygon_count = 8 * m_lineCount;
+            index_count = 3 * polygon_count;
+            vertex_count = 6 * m_lineCount;
+        } else {
+            captainslog_assert(0);
+        }
+
+        DynamicIBAccessClass index_buffer(
+            sort ? IndexBufferClass::BUFFER_TYPE_DYNAMIC_SORTING : IndexBufferClass::BUFFER_TYPE_DYNAMIC_DX8, index_count);
+        { // added to control the lifetime of the WriteLockClass object
+            DynamicIBAccessClass::WriteLockClass index_lock(&index_buffer);
+
+            unsigned short *index = index_lock.Get_Index_Array();
+
+            if (m_mode == TETRAHEDRON) {
+                for (int i = 0; i < m_lineCount; ++i) {
+                    *index++ = short(4 * i + 0);
```

Comment:
i think it would help to split this up into the polygons. So just put an additional return after each 3 so that you end up with 4 polygons as per `Get_Polygon_Count` same goes for `PRISM`